File: aihwbench/comparability.py
"""Comparison safety classification.

Classifies whether two benchmark results are safe to compare:

- STRICTLY_COMPARABLE: all materially relevant dimensions match.
- CONDITIONALLY_COMPARABLE: model/workload match but caveats exist.
- NOT_COMPARABLE: direct metric comparison would be misleading.

Never decides winners; only decides whether a comparison is safe.
Machine-readable reasons are included for automation/CI.
"""

from __future__ import annotations

from typing import Any

STRICTLY_COMPARABLE = "STRICTLY_COMPARABLE"
CONDITIONALLY_COMPARABLE = "CONDITIONALLY_COMPARABLE"
NOT_COMPARABLE = "NOT_COMPARABLE"

# Differing strictly => NOT_COMPARABLE.
_STRICT = (
    "model.name",
    "model.checksum",
    "model.format",
    "model.quantization",
    "model.revision",
    "model.tokenizer",
    "runtime.name",
    "runtime.backend",
    "runtime.device",
    "reproducibility.workload_type",
    "reproducibility.prompt",
    "reproducibility.max_tokens",
    "reproducibility.temperature",
    "reproducibility.seed",
    "reproducibility.context_length",
    "reproducibility.batch_size",
    "reproducibility.concurrency",
    "reproducibility.warmup_runs",
    "reproducibility.iterations",
)

# Differing these => CONDITIONALLY_COMPARABLE.
_CONDITIONAL = (
    "reproducibility.power_profile",
    "system.os_version",
    "runtime.version",
)


def _get(a: dict[str, Any], path: str) -> Any:
    value: Any = a
    for part in path.split("."):
        if not isinstance(value, dict):
            return None
        value = value.get(part)
        if value is None:
            return None
    return value


def _same(a: Any, b: Any) -> bool:
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    return bool(a == b)


def comparability_warnings(a: dict[str, Any], b: dict[str, Any]) -> list[str]:
    """Human-readable warnings when results are not strictly comparable."""
    reasons: list[str] = compare_classification(a, b)["reasons"]
    return reasons


def compare_classification(a: dict[str, Any], b: dict[str, Any]) -> dict[str, Any]:
    """Classify comparability of two result documents.

    Returns dict with "classification", "reasons" (human), and
    "machine_reasons" (stable dotted keys).
    """
    reasons: list[str] = []
    machine: list[str] = []

    strict_diffs = [p for p in _STRICT if not _same(_get(a, p), _get(b, p))]
    conditional_diffs = [p for p in _CONDITIONAL if not _same(_get(a, p), _get(b, p))]

    for path in strict_diffs:
        if _get(a, path) is None and _get(b, path) is None:
            continue
        va, vb = _get(a, path), _get(b, path)
        if path == "model.name":
            reasons.append(f"models differ: {va!r} vs {vb!r}")
        else:
            reasons.append(f"{path}: {va!r} vs {vb!r}")
        machine.append(path)

    for path in conditional_diffs:
        if _get(a, path) is None and _get(b, path) is None:
            continue
        reasons.append(
            f"{path} differs ({_get(a, path)!r} vs {_get(b, path)!r}) - interpret as conditional"
        )
        machine.append(path)

    if _get(a, "system.cpu") != _get(b, "system.cpu") or _get(a, "system.gpu") != _get(
        b, "system.gpu"
    ):
        reasons.append("hardware differs between results - treat as cross-platform reference only")
        machine.append("system.hardware")

    if strict_diffs:
        classification = NOT_COMPARABLE
    elif machine:
        classification = CONDITIONALLY_COMPARABLE
    else:
        classification = STRICTLY_COMPARABLE

    return {
        "classification": classification,
        "reasons": reasons,
        "machine_reasons": sorted(set(machine)),
    }


def assert_comparable(a: dict[str, Any], b: dict[str, Any]) -> None:
    """Raise ValueError when two results are NOT_COMPARABLE."""
    result = compare_classification(a, b)
    if result["classification"] == NOT_COMPARABLE:
        detail = "\n  - ".join(result["reasons"])
        raise ValueError("Results are NOT_COMPARABLE:\n  - " + detail)

File: web/src/lib/comparability.ts
// Browser-side comparison-safety classifier.
//
// `aihwbench/comparability.py` stays canonical. The dashboard needs the same
// verdict for any pair a reader selects, and the pair is chosen at runtime, so
// it cannot be precomputed — the logic has to exist here too.
//
// The rule tables are not duplicated: they are generated into
// `data/comparability.json` from the Python module, which also carries
// reference verdicts that the tests replay through this implementation. A
// change to either side that alters a verdict fails the build.
//
// The classifier never picks a winner. It decides only whether a comparison
// may honestly be made.

import type { BenchmarkResultDoc, ComparabilityRules } from './types'

export const STRICTLY_COMPARABLE = 'STRICTLY_COMPARABLE'
export const CONDITIONALLY_COMPARABLE = 'CONDITIONALLY_COMPARABLE'
export const NOT_COMPARABLE = 'NOT_COMPARABLE'

export type Verdict =
  | typeof STRICTLY_COMPARABLE
  | typeof CONDITIONALLY_COMPARABLE
  | typeof NOT_COMPARABLE

export interface Classification {
  classification: Verdict
  reasons: string[]
  machine_reasons: string[]
}

/** Read a dotted path, returning undefined rather than throwing. */
function get(doc: unknown, path: string): unknown {
  let value: unknown = doc
  for (const part of path.split('.')) {
    if (value === null || typeof value !== 'object') return undefined
    value = (value as Record<string, unknown>)[part]
  }
  return value ?? undefined
}

function isMissing(value: unknown): boolean {
  return value === undefined || value === null
}

/**
 * Two absent values agree; one absent and one present do not.
 *
 * Mirrors `_same` in the Python module, including the deliberate asymmetry:
 * a field neither result recorded is genuine agreement about an optional
 * field, but a field only one recorded is a difference.
 */
function same(a: unknown, b: unknown): boolean {
  if (isMissing(a) && isMissing(b)) return true
  if (isMissing(a) || isMissing(b)) return false
  return a === b
}

export function classify(
  rules: ComparabilityRules,
  a: BenchmarkResultDoc | undefined,
  b: BenchmarkResultDoc | undefined,
): Classification {
  const reasons: string[] = []
  const machine: string[] = []

  if (!a || !b) {
    return {
      classification: NOT_COMPARABLE,
      reasons: ['two results are required for a comparison'],
      machine_reasons: [rules.insufficient_metadata_reason],
    }
  }

  // Rule 1: required provenance must be present on both sides. Absence of
  // evidence is not evidence of sameness.
  const missing = rules.required_present.filter(
    (path) => isMissing(get(a, path)) || isMissing(get(b, path)),
  )

  const strictDiffs = rules.strict.filter((path) => !same(get(a, path), get(b, path)))
  const conditionalDiffs = rules.conditional.filter(
    (path) => !same(get(a, path), get(b, path)),
  )

  for (const path of strictDiffs) {
    if (isMissing(get(a, path)) && isMissing(get(b, path))) continue
    const va = get(a, path)
    const vb = get(b, path)
    reasons.push(
      path === 'model.name'
        ? `models differ: ${String(va)} vs ${String(vb)}`
        : `${path}: ${String(va)} vs ${String(vb)}`,
    )
    machine.push(path)
  }

  for (const path of conditionalDiffs) {
    if (isMissing(get(a, path)) && isMissing(get(b, path))) continue
    reasons.push(
      `${path} differs (${String(get(a, path))} vs ${String(get(b, path))}) - interpret as conditional`,
    )
    machine.push(path)
  }

  if (get(a, 'system.cpu') !== get(b, 'system.cpu') || get(a, 'system.gpu') !== get(b, 'system.gpu')) {
    reasons.push('hardware differs between results - treat as cross-platform reference only')
    machine.push('system.hardware')
  }

  if (missing.length > 0) {
    reasons.push(
      `required provenance is missing, so these results cannot be compared: ${missing.join(', ')}`,
    )
    machine.push(rules.insufficient_metadata_reason)
  }

  let classification: Verdict
  if (missing.length > 0 || strictDiffs.length > 0) {
    classification = NOT_COMPARABLE
  } else if (machine.length > 0) {
    classification = CONDITIONALLY_COMPARABLE
  } else {
    classification = STRICTLY_COMPARABLE
  }

  return {
    classification,
    reasons,
    machine_reasons: Array.from(new Set(machine)).sort(),
  }
}

/** Human-readable one-liner for a verdict badge. */
export function verdictSummary(verdict: Verdict): string {
  switch (verdict) {
    case STRICTLY_COMPARABLE:
      return 'Every materially relevant dimension matches. These numbers may be compared directly.'
    case CONDITIONALLY_COMPARABLE:
      return 'The experiment matches, but something worth stating differs. Read the deltas with the caveats below.'
    default:
      return 'These runs measured different things, or did not record enough to tell. A direct comparison would mislead.'
  }
}
